File: frontend/lib/account.ts
const BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"

async function authRequest<T>(
  path: string,
  token: string,
  init?: RequestInit,
): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
      ...(init?.headers ?? {}),
    },
  })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    const detail = body?.detail
    const message =
      typeof detail === "string"
        ? detail
        : detail?.code ?? JSON.stringify(detail) ?? `HTTP ${res.status}`
    throw new Error(message)
  }
  return res.json() as Promise<T>
}

export interface ExportJob {
  id: string
  status: "pending" | "processing" | "ready" | "failed"
  presigned_url: string | null
  presigned_url_expires_at: string | null
  error: string | null
  created_at: string
  completed_at: string | null
}

export interface ExportListItem {
  id: string
  status: string
  presigned_url: string | null
  presigned_url_expires_at: string | null
  created_at: string
  completed_at: string | null
}

export async function startExport(token: string): Promise<{ job_id: string }> {
  return authRequest("/api/account/export", token, { method: "POST" })
}

export async function getExportJob(
  token: string,
  jobId: string,
): Promise<ExportJob> {
  return authRequest(`/api/account/export/${jobId}`, token)
}

export async function listExports(token: string): Promise<ExportListItem[]> {
  return authRequest("/api/account/exports", token)
}

export async function closeAccount(
  token: string,
  cancelSubscription = true,
): Promise<{ ok: boolean; scheduled_delete_at: string }> {
  return authRequest("/api/account/close", token, {
    method: "POST",
    body: JSON.stringify({ cancel_subscription: cancelSubscription }),
  })
}

export async function cancelAccountClosure(
  token: string,
): Promise<{ ok: boolean }> {
  return authRequest("/api/account/close/cancel", token, { method: "POST" })
}

export async function patchDisplayName(
  token: string,
  displayName: string,
): Promise<{ ok: boolean; display_name: string }> {
  return authRequest("/api/account/profile", token, {
    method: "PATCH",
    body: JSON.stringify({ display_name: displayName }),
  })
}

export async function sendEmailVerification(token: string): Promise<void> {
  await authRequest("/api/auth/verify/send", token, { method: "POST" })
}

export function pollExportUntilReady(
  token: string,
  jobId: string,
  intervalMs = 5000,
  maxAttempts = 60,
): { promise: Promise<ExportJob>; cancel: () => void } {
  let cancelled = false
  let timer: ReturnType<typeof setTimeout> | null = null

  const promise = new Promise<ExportJob>((resolve, reject) => {
    let attempts = 0

    const tick = async () => {
      if (cancelled) return
      attempts += 1
      try {
        const job = await getExportJob(token, jobId)
        if (job.status === "ready") {
          resolve(job)
          return
        }
        if (job.status === "failed") {
          reject(new Error(job.error ?? "Export failed"))
          return
        }
        if (attempts >= maxAttempts) {
          reject(new Error("Export timed out"))
          return
        }
        timer = setTimeout(tick, intervalMs)
      } catch (e) {
        reject(e)
      }
    }

    void tick()
  })

  const cancel = () => {
    cancelled = true
    if (timer) clearTimeout(timer)
  }

  return { promise, cancel }
}

File: backend/app/routers/account.py
"""Account export and closure routes (IMPLEMENTATION_PLAN §6)."""

from __future__ import annotations

import uuid
from datetime import datetime, timedelta, timezone
from typing import Annotated, Any

import structlog
from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Request, Security, status
from fastapi.security import APIKeyHeader
from pydantic import BaseModel, Field
from slowapi.util import get_remote_address
from sqlalchemy import desc, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.db.engine import async_session_factory, get_db
from app.limiter import limiter
from app.models.export import ExportJob, ExportJobStatus
from app.models.export import ClosureRequest
from app.models.user import User
from app.services.auth.dependencies import get_current_user
from app.services.auth.tokens import decode_access_token
from app.services.export.assembler import process_export_job
from app.services.export.closure import cancel_closure, run_closure_tick, schedule_closure

log = structlog.get_logger("account.router")

router = APIRouter(prefix="/api/account", tags=["account"])

_scheduler_header = APIKeyHeader(name="X-Scheduler-Secret", auto_error=False)


def _rate_limit_user_key(request: Request) -> str:
    auth = request.headers.get("Authorization", "")
    if auth.lower().startswith("bearer "):
        token = auth[7:].strip()
        try:
            claims = decode_access_token(token, expected_type="access")
            subject = str(claims.get("sub") or "").strip()
            if subject:
                return f"user:{subject}"
        except Exception:  # noqa: BLE001
            pass
        return f"token:{token[:64]}"
    return get_remote_address(request)


class ExportCreateResponse(BaseModel):
    job_id: uuid.UUID


class ExportJobResponse(BaseModel):
    id: uuid.UUID
    status: str
    presigned_url: str | None = None
    presigned_url_expires_at: datetime | None = None
    error: str | None = None
    created_at: datetime
    completed_at: datetime | None = None


class ExportListItem(BaseModel):
    id: uuid.UUID
    status: str
    presigned_url: str | None = None
    presigned_url_expires_at: datetime | None = None
    created_at: datetime
    completed_at: datetime | None = None


class CloseAccountRequest(BaseModel):
    cancel_subscription: bool = True


class CloseAccountResponse(BaseModel):
    ok: bool
    scheduled_delete_at: datetime


class ProfilePatchRequest(BaseModel):
    display_name: str = Field(..., min_length=1, max_length=200)


async def _run_export_background(job_id: uuid.UUID) -> None:
    async with async_session_factory() as session:
        try:
            await process_export_job(session, job_id)
            await session.commit()
        except Exception as exc:  # noqa: BLE001
            await session.rollback()
            log.error("export.background_failed", job_id=str(job_id), error=str(exc))


def _job_response(job: ExportJob) -> ExportJobResponse:
    return ExportJobResponse(
        id=job.id,
        status=job.status.value,
        presigned_url=job.presigned_url,
        presigned_url_expires_at=job.presigned_url_expires_at,
        error=job.error,
        created_at=job.created_at,
        completed_at=job.completed_at,
    )


@router.patch("/profile")
@limiter.limit("30/minute")
async def patch_profile(
    request: Request,
    body: ProfilePatchRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> dict[str, Any]:
    user.display_name = body.display_name.strip()
    await db.flush()
    return {"ok": True, "display_name": user.display_name}


@router.post("/export", status_code=202)
@limiter.limit("2/day", key_func=_rate_limit_user_key)
async def create_export(
    request: Request,
    background_tasks: BackgroundTasks,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ExportCreateResponse:
    cutoff = datetime.now(timezone.utc) - timedelta(hours=24)
    recent_count = (
        await db.execute(
            select(func.count())
            .select_from(ExportJob)
            .where(ExportJob.user_id == user.id)
            .where(ExportJob.created_at >= cutoff)
        )
    ).scalar_one()
    if recent_count >= 2:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail={"code": "export_rate_limited", "limit": "2/24h"},
        )

    job = ExportJob(user_id=user.id, status=ExportJobStatus.pending)
    db.add(job)
    await db.flush()
    background_tasks.add_task(_run_export_background, job.id)
    return ExportCreateResponse(job_id=job.id)


@router.get("/export/{job_id}")
@limiter.limit("120/minute")
async def get_export(
    request: Request,
    job_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ExportJobResponse:
    job = (
        await db.execute(
            select(ExportJob).where(
                ExportJob.id == job_id,
                ExportJob.user_id == user.id,
            )
        )
    ).scalar_one_or_none()
    if job is None:
        raise HTTPException(status_code=404, detail="Export job not found")
    return _job_response(job)


@router.get("/exports")
@limiter.limit("30/minute")
async def list_exports(
    request: Request,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> list[ExportListItem]:
    cutoff = datetime.now(timezone.utc) - timedelta(days=30)
    rows = (
        await db.execute(
            select(ExportJob)
            .where(
                ExportJob.user_id == user.id,
                ExportJob.created_at >= cutoff,
            )
            .order_by(desc(ExportJob.created_at))
        )
    ).scalars().all()
    return [
        ExportListItem(
            id=j.id,
            status=j.status.value,
            presigned_url=j.presigned_url,
            presigned_url_expires_at=j.presigned_url_expires_at,
            created_at=j.created_at,
            completed_at=j.completed_at,
        )
        for j in rows
    ]


@router.post("/close")
@limiter.limit("5/day", key_func=_rate_limit_user_key)
async def close_account(
    request: Request,
    body: CloseAccountRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> CloseAccountResponse:
    if user.is_closure_pending:
        existing = (
            await db.execute(
                select(ClosureRequest).where(
                    ClosureRequest.user_id == user.id,
                    ClosureRequest.cancelled_at.is_(None),
                )
            )
        ).scalar_one_or_none()
        scheduled = (
            existing.scheduled_delete_at
            if existing
            else user.closure_requested_at + timedelta(days=settings.ACCOUNT_CLOSURE_GRACE_DAYS)  # type: ignore[operator]
        )
        return CloseAccountResponse(ok=True, scheduled_delete_at=scheduled)

    row = await schedule_closure(
        db,
        user=user,
        cancel_subscription=body.cancel_subscription,
    )
    return CloseAccountResponse(ok=True, scheduled_delete_at=row.scheduled_delete_at)


@router.post("/close/cancel")
@limiter.limit("5/day", key_func=_rate_limit_user_key)
async def cancel_account_closure(
    request: Request,
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> dict[str, bool]:
    ok = await cancel_closure(db, user=user)
    if not ok:
        raise HTTPException(status_code=404, detail="No pending closure request")
    return {"ok": True}


@router.delete("", status_code=200)
async def scheduler_delete_due_accounts(
    request: Request,
    db: Annotated[AsyncSession, Depends(get_db)],
    x_scheduler_secret: Annotated[str | None, Security(_scheduler_header)] = None,
) -> dict[str, Any]:
    """Internal endpoint for EventBridge closure tick."""
    secret = settings.INTERNAL_SCHEDULER_SECRET
    if not secret or x_scheduler_secret != secret:
        raise HTTPException(status_code=401, detail="Unauthorized")
    result = await run_closure_tick(db)
    return {
        "ok": True,
        "inspected": result.inspected,
        "deleted": [str(uid) for uid in result.deleted],
        "reminders_sent": result.reminders_sent,
    }


__all__ = ["router"]
